packages/bun-npm/src/install.ts

```diff
@@ -0,0 +1,151 @@
+    if (exitCode === 0) {
+      return exe;
+    }
+    throw new Error(stderr || stdout);
+  }
+  try {
+    return resolveBun();
+  } catch (error) {
+    console.debug("resolveBun failed", error);
+    console.error(
+      `Failed to find package "${npmPackage}".`,
+      `You may have used the "--no-optional" flag when running "npm install".`,
+    );
+  }
+  const cwd = join("node_modules", npmPackage);
+  try {
+    installBun(platform, cwd);
+  } catch (error) {
+    console.debug("installBun failed", error);
+    console.error(
+      `Failed to install package "${npmPackage}" using "npm install".`,
+      error,
+    );
+    try {
+      await downloadBun(platform, cwd);
+    } catch (error) {
+      console.debug("downloadBun failed", error);
+      console.error(
+        `Failed to download package "${npmPackage}" from "registry.npmjs.org".`,
+        error,
+      );
+    }
+  }
+  return resolveBun();
+}
+
+function installBun(platform: Platform, dst: string): void {
+  const npmPackage = `${npmOwner}/${platform.bin}`;
+  const cwd = tmp();
+  try {
+    write(join(cwd, "package.json"), "{}");
+    const { exitCode } = spawn(
+      "npm",
+      [
+        "install",
+        "--loglevel=error",
+        "--prefer-offline",
+        "--no-audit",
+        "--progress=false",
+        `${npmPackage}@${npmVersion}`,
+      ],
+      {
+        cwd,
+        stdio: "pipe",
+        env: {
+          ...process.env,
+          npm_config_global: undefined,
+        },
+      },
+    );
+    if (exitCode === 0) {
+      rename(join(cwd, "node_modules", npmPackage), dst);
+    }
+  } finally {
+    try {
+      rm(cwd);
+    } catch (error) {
+      console.debug("rm failed", error);
+      // There is nothing to do if the directory cannot be cleaned up.
+    }
+  }
+}
+
+async function downloadBun(platform: Platform, dst: string): Promise<void> {
+  const response = await fetch(
+    `https://registry.npmjs.org/${npmOwner}/${platform.bin}/-/${platform.bin}-${npmVersion}.tgz`,
+  );
+  const tgz = await response.arrayBuffer();
+  let buffer: Buffer;
+  try {
+    buffer = unzipSync(tgz);
+  } catch (cause) {
+    throw new Error("Invalid gzip data", { cause });
+  }
+  function str(i: number, n: number): string {
+    return String.fromCharCode(...buffer.subarray(i, i + n)).replace(
+      /\0.*$/,
+      "",
+    );
+  }
+  let offset = 0;
+  while (offset < buffer.length) {
+    const name = str(offset, 100).replace("package/", "");
+    const size = parseInt(str(offset + 124, 12), 8);
+    offset += 512;
+    if (!isNaN(size)) {
+      write(join(dst, name), buffer.subarray(offset, offset + size));
+      if (name === platform.exe) {
+        try {
+          chmod(join(dst, name), 0o755);
+        } catch (error) {
+          console.debug("chmod failed", error);
+        }
+      }
+      offset += (size + 511) & ~511;
+    }
+  }
+}
+
+export function optimizeBun(path: string): void {
```

Comment:
Let's hard fail if they don't successfully symlink `bun` binary to not start instances of `bun` via `node`

can print a message saying something like "Your package manager doesn't seem to support linking bun install. To use bun, please install via curl https://bun.sh/install | bash"

Reply:
f52831ba420e6dffb2feed74c4c0d3185ce9b155